impl/src/context.rs

```diff
@@ -53,6 +56,137 @@ pub struct SourceDetails {
   pub git_data: Option<GitRepo>,
 }
 
+#[derive(Debug, Clone)]
+pub struct SourceId {
+  pub kind: SourceKind,
+  pub url: Url,
+  pub precise: Option<String>,
+}
+
+#[derive(Debug, Clone)]
+pub enum SourceKind {
```

Comment:
Can you run rustfmt over this file? Looks like some of the file is using the proper 4 space scheme instead of my terrible 2 space scheme.

I really should fix that.